### PR TITLE
support olap writeback for cassandra and rocksdb

### PR DESCRIPTION
**Pull request by zhoney**
Change-Id: I947c3a96eb7aa67ee9eff5f7fb248e9f4539e91b

### CODE REVIEW

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 1]
```diff
@@ -258,6 +284,70 @@ public Id rebuildIndex(SchemaElement schema, Set<Id> dependencies) {
...
+        if (propertyKey.readFrequency() == ReadFrequency.OLTP) {
+            return null;
+        }
+
+        Id taskId = this.createOlapPk(propertyKey);
+        if (propertyKey.readFrequency() == ReadFrequency.OLAP_NONE) {
+            return taskId;
+        }
+
+        String indexName = ALL +"_by_" + propertyKey.name();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java:297]: space

File: hugegraph-core/src/main/java/com/baidu/hugegraph/job/schema/EdgeLabelRemoveCallable.java [hunk 2]
```diff
@@ -43,7 +43,7 @@ public Object execute() {
         return null;
     }
 
-    protected static void removeEdgeLabel(HugeGraphParams graph, Id id) {
+    private static void removeEdgeLabel(HugeGraphParams graph, Id id) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/job/schema/EdgeLabelRemoveCallable.java:46]: keep protected

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java [hunk 3]
```diff
@@ -38,10 +38,12 @@
                                                Cloneable {
 
     public static final int MAX_PRIMITIVE_SYS_ID = 32;
-    public static final int NEXT_PRIMITIVE_SYS_ID = 7;
+    public static final int NEXT_PRIMITIVE_SYS_ID = 8;
 
     public static final Id NONE_ID = IdGenerator.ZERO;
+    public static final Id ALL_ID = IdGenerator.of(-7);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java:44]: ALL means what

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 4]
```diff
@@ -80,14 +83,15 @@
     private final String database;
 
     private final BackendStoreProvider provider;
-    private final Map<HugeType, RocksDBTable> tables;
+    private final Map<String, RocksDBTable> tables;
 
     private String dataPath;
-    private RocksDBSessions sessions;
+    protected RocksDBSessions sessions;
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:89]: keep private

File: hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java [hunk 5]
```diff
@@ -262,9 +285,23 @@ private void mutate(CassandraSessionPool.Session session,
...
-
-        CassandraTable table = this.table(CassandraTable.tableType(query));
-        return table.query(this.sessions.session(), query);
+        HugeType type = CassandraTable.tableType(query);
+        String tableName = query.olap() ?
+                           this.olapTableName(type) : type.string();
+        CassandraTable table = this.table(tableName);
+        Iterator<BackendEntry> entrys = table.query(this.session(), query);
+        Set<Id> olapPks = query.olapPks();
+        String graphStore = this.conf.get(CoreOptions.STORE_GRAPH);
```
Comment (javeme) [inline hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java:294]: remove this.conf and define graphStore as field

File: hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java [hunk 6]
```diff
@@ -262,9 +285,23 @@ private void mutate(CassandraSessionPool.Session session,
...
     public Iterator<BackendEntry> query(Query query) {
         this.checkOpened();
-
-        CassandraTable table = this.table(CassandraTable.tableType(query));
-        return table.query(this.sessions.session(), query);
+        HugeType type = CassandraTable.tableType(query);
+        String tableName = query.olap() ?
+                           this.olapTableName(type) : type.string();
+        CassandraTable table = this.table(tableName);
+        Iterator<BackendEntry> entrys = table.query(this.session(), query);
```
Comment (javeme) [inline hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java:292]: entries

File: hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java [hunk 7]
```diff
@@ -262,9 +285,23 @@ private void mutate(CassandraSessionPool.Session session,
...
+        Set<Id> olapPks = query.olapPks();
```
Comment (javeme) [inline hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java:293]: add comment: Merge olap results as needed

File: hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java [hunk 8]
```diff
@@ -528,7 +565,11 @@ protected void clearTables() {
     protected void truncateTables() {
         CassandraSessionPool.Session session = this.sessions.session();
         for (CassandraTable table : this.tables()) {
-            table.truncate(session);
+            if (table.isOlap()) {
+                table.dropTable(session);
```
Comment (javeme) [inline hugegraph-cassandra/src/main/java/com/baidu/hugegraph/backend/store/cassandra/CassandraStore.java:569]: let table itself do drop in truncate() method

File: hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java [hunk 9]
```diff
@@ -255,6 +257,13 @@ public void serverStarted(Id serverId, NodeRole serverRole) {
                  serverId, serverRole, this.name);
         this.serverInfoManager().initServerInfo(serverId, serverRole);
 
+        LOG.info("Search olap property key for graph '{}'", this.name);
+        for (PropertyKey pk : this.schemaTransaction().getPropertyKeys()) {
+            if (pk.readFrequency().olap()) {
+                this.schemaTransaction().initAndRegisterOlapTable(pk.id());
+            }
+        }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java:265]: move to this.schemaTransaction()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java [hunk 10]
```diff
@@ -88,6 +92,8 @@ public Query(HugeType resultType, Query originQuery) {
 
         this.aggregate = null;
         this.showExpired = false;
+        this.olap = false;
+        this.olapPks = new IdSet(CollectionType.EC);
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java:96]: also replace conditions and ids to EC collection

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java [hunk 11]
```diff
@@ -343,6 +350,26 @@ public boolean paging() {
         return this.page != null;
     }
 
+    public void olap(boolean olap) {
+        this.olap = olap;
+    }
+
+    public boolean olap() {
+        return this.olap;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java:358]: can return olapPks != null && olapPks not empty

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java [hunk 12]
```diff
@@ -343,6 +350,26 @@ public boolean paging() {
...
+    }
+
+    public void olapPks(Set<Id> olapPks) {
+        this.olapPks = olapPks;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java:362]: can init with null to lazy set

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 13]
```diff
@@ -1116,6 +1135,7 @@ private static ConditionQuery constructQuery(ConditionQuery query,
                 joinedValues = escapeIndexValueIfNeeded(joinedValues);
 
                 indexQuery = new ConditionQuery(indexType.type(), query);
+                indexQuery.olap(indexLabel.olap());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java:1138]: move to copyBasic()

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 14]
```diff
@@ -153,6 +156,17 @@ public void updateLabelIndex(HugeElement element, boolean removed) {
...
     public void updateVertexIndex(HugeVertex vertex, boolean removed) {
+        if (vertex.olap()) {
+            Id pkId = vertex.getProperties().keySet().iterator().next();
+            List<IndexLabel> indexLabels = this.params().schemaTransaction()
+                                               .getIndexLabels();
+            for (IndexLabel il : indexLabels) {
+                if (il.indexFields().contains(pkId)) {
+                    this.updateIndex(il.id(), vertex, removed);
+                }
+            }
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java:167]: add a method

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java [hunk 15]
```diff
@@ -153,6 +156,17 @@ public void updateLabelIndex(HugeElement element, boolean removed) {
 
     @Watched(prefix = "index")
     public void updateVertexIndex(HugeVertex vertex, boolean removed) {
+        if (vertex.olap()) {
+            Id pkId = vertex.getProperties().keySet().iterator().next();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java:160]: only one pkId?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java (hunk 3 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java:44]: ALL means all vertex labels

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java (hunk 15 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/GraphIndexTransaction.java:160]: yes. write back only one olap property for one vertex id

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java (hunk 11 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java:358]: olapPks is for query Vertex, olap is for index query

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java (hunk 12 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java:362]: olapPks is for query Vertex, olap is for index query

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java (hunk 10 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/query/Query.java:96]: ids need keep order, using LinkedHashSet, can't be replaced

**Comment by javeme** [comment]
```
Error:  Tests run: 371, Failures: 1, Errors: 0, Skipped: 1, Time elapsed: 628.431 s <<< FAILURE! - in com.baidu.hugegraph.unit.UnitTestSuite
Error:  testToString(com.baidu.hugegraph.unit.core.QueryTest)  Time elapsed: 0.013 s  <<< FAILURE!
org.junit.ComparisonFailure: expected:<...e id in [1, 3] and [[LABEL == 3, PROPERTIES < 18]]`> but was:<...e id in [1, 3] and [[PROPERTIES < 18, LABEL == 3]]`>
	at com.baidu.hugegraph.unit.core.QueryTest.testToString(QueryTest.java:120)
```

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java (hunk 3 above)
Comment (javeme) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java:44]: still not get it

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java [hunk 16]
```diff
@@ -165,4 +167,57 @@ public static int schemaId(Id id) {
...
+                         "Expect index label, but actual schema type is " +
+                         "'%s'", this.schemaElement.getClass());
+            return (IndexLabel) this.schemaElement;
+        }
+
+        public SchemaElement schemaElement() {
+            return this.schemaElement;
+        }
+
+        public void task(Id task) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java:215]: unused?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java [hunk 17]
```diff
@@ -165,4 +167,57 @@ public static int schemaId(Id id) {
...
+        }
+
+        public IndexLabel indexLabel() {
+            E.checkState(this.schemaElement instanceof IndexLabel,
+                         "Expect index label, but actual schema type is " +
+                         "'%s'", this.schemaElement.getClass());
+            return (IndexLabel) this.schemaElement;
+        }
+
+        public SchemaElement schemaElement() {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java:211]: unused?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java [hunk 18]
```diff
@@ -165,4 +167,57 @@ public static int schemaId(Id id) {
...
+            E.checkNotNull(schemaElement, "schema element");
+            this.schemaElement = schemaElement;
+        }
+
+        public void propertyKey(PropertyKey propertyKey) {
+            E.checkNotNull(schemaElement, "property key");
+            this.schemaElement = propertyKey;
+        }
+
+        public void indexLabel(IndexLabel indexLabel) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java:192]: not covered by tests

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java [hunk 19]
```diff
@@ -19,6 +19,9 @@
 
 package com.baidu.hugegraph.schema.builder;
 
+import static com.baidu.hugegraph.schema.SchemaElement.ALL;
+import static com.baidu.hugegraph.schema.VertexLabel.ALL_VL;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/IndexLabelBuilder.java:23]: don't import static as possible

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/PropertyKeyBuilder.java [hunk 20]
```diff
@@ -425,5 +457,13 @@ private void checkOlap() {
                       "Not allow to set aggregate type '%s' for olap " +
                       "property key '%s'", this.aggregateType, this.name);
         }
+
+        if (this.readFrequency == ReadFrequency.OLAP_RANGE &&
+            !this.dataType.isNumber() && !this.dataType.isDate()) {
+            throw new NotAllowException(
+                      "Not allow to set read frequency to OLAP_RANGE for " +
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/schema/builder/PropertyKeyBuilder.java:464]: Not allowed

File: hugegraph-core/src/main/java/com/baidu/hugegraph/type/define/SchemaStatus.java [hunk 21]
```diff
@@ -27,11 +27,13 @@
...
-    DELETING(4, "deleting"),
+    CLEARING(4, "clearing"),
 
-    UNDELETED(5, "undeleted"),
+    DELETING(5, "deleting"),
 
-    INVALID(6, "invalid");
+    UNDELETED(6, "undeleted"),
+
+    INVALID(7, "invalid");
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/type/define/SchemaStatus.java:36]: keep the original code to keep compatible

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBFeatures.java [hunk 22]
```diff
@@ -138,6 +138,6 @@ public boolean supportsTtl() {
 
     @Override
     public boolean supportsOlapProperties() {
-        return false;
+        return true;
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBFeatures.java:141]: need to bump backend version?

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 23]
```diff
@@ -134,15 +138,27 @@ private void registerMetaHandlers() {
...
         assert type != null;
-        RocksDBTable table = this.tables.get(type);
+        return this.table(type.string());
+    }
+
+    protected final RocksDBTable table(String name) {
+        RocksDBTable table = this.tables.get(name);
         if (table == null) {
-            throw new BackendException("Unsupported table type: %s", type);
+            throw new BackendException("Unsupported table: %s", name);
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:161]: '%s'

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 24]
```diff
@@ -404,19 +422,28 @@ public void mutate(BackendMutation mutation) {
...
-                table.insert(session, entry);
+                name = entry.type() == HugeType.OLAP ?
+                       this.olapTableName(entry.subId()) :
+                       entry.type().string();
+                this.table(name).insert(session, entry);
                 break;
             case DELETE:
-                table.delete(session, entry);
+                name = entry.olap() ?
+                       this.olapTableName(entry.type()) : entry.type().string();
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:437]: wrap line after ":"

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 25]
```diff
@@ -404,19 +422,28 @@ public void mutate(BackendMutation mutation) {
...
+                this.table(name).delete(session, entry);
                 break;
             case APPEND:
-                table.append(session, entry);
+                name = entry.olap() ?
+                       this.olapTableName(entry.type()) : entry.type().string();
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:442]: ditto

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 26]
```diff
@@ -431,10 +458,26 @@ private void mutate(Session session, BackendAction item) {
         readLock.lock();
         try {
             this.checkOpened();
-
             HugeType tableType = RocksDBTable.tableType(query);
-            RocksDBTable table = this.table(tableType);
-            return table.query(this.session(tableType), query);
+
+            String tableName = query.olap() ? this.olapTableName(tableType) :
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:463]: ditto

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 27]
```diff
@@ -941,5 +995,43 @@ public long getCounter(HugeType type) {
...
+
+        @Override
+        public void createOlapTable(Id pkId) {
+            RocksDBTable table = new RocksDBTables.Olap(this.store(), pkId);
+            this.createTable(super.sessions, table.table());
+            registerTableManager(this.olapTableName(pkId), table);
+        }
+
+        @Override
+        public void checkAndRegisterOlapTable(Id pkId) {
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:1007]: not covered by tests

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java [hunk 28]
```diff
@@ -336,4 +336,88 @@ public ShardIndex(String database) {
             super(database, TABLE);
         }
     }
+
+    public static class Olap extends RocksDBTable {
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java:340]: OlapTable

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java [hunk 29]
```diff
@@ -336,4 +336,88 @@ public ShardIndex(String database) {
...
+        }
+
+        protected OlapSecondaryIndex(String store, String table) {
+            super(joinTableName(store, table));
+        }
+    }
+
+    public static class OlapRangeIntIndex extends RangeIntIndex {
+
+        public static final String TABLE = HugeType.OLAP.string();
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java:374]: is it the same table name as OlapSecondaryIndex

File: hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java [hunk 30]
```diff
@@ -834,7 +833,8 @@ public boolean existsEdgeLabel(String name) {
 
     @Override
     public void addIndexLabel(SchemaLabel schemaLabel, IndexLabel indexLabel) {
-        assert this.name.equals(schemaLabel.graph().name());
+        assert ALL_VL.equals(schemaLabel) ||
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java:856]: ALL_VL means all of vertex labels

File: hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java [hunk 31]
```diff
@@ -410,6 +412,17 @@ public void resumeSnapshot() {
         LOG.info("Graph '{}' has resumed from snapshot", this.name);
     }
 
+    private void clearVertexCache() {
+        Future<?> future = this.graphEventHub.notify(Events.CACHE, "clear",
+                                                     HugeType.VERTEX);
```
[the remark this answers is not in the document]
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/StandardHugeGraph.java:417]: olap properties only affect vertex query results, seems no need to clear edge cache

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java (hunk 16 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java:215]: yes. deleted.

File: hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java (hunk 17 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/schema/SchemaElement.java:211]: yes. deleted.

File: hugegraph-core/src/main/java/com/baidu/hugegraph/type/define/SchemaStatus.java (hunk 21 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/type/define/SchemaStatus.java:36]: done

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java (hunk 29 above)
Comment (zhoney) [reply hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBTables.java:374]: no, it's like g_ap_ii

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 32]
```diff
@@ -185,8 +181,8 @@ public void addVertexLabel(VertexLabel vertexLabel) {
     @Watched(prefix = "schema")
     public VertexLabel getVertexLabel(Id id) {
         E.checkArgumentNotNull(id, "Vertex label id can't be null");
-        if (ALL_ID.equals(id)) {
-            return ALL_VL;
+        if (SchemaElement.ALL_ID.equals(id)) {
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java:184]: OLAP_VL_ID

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java [hunk 33]
```diff
@@ -185,8 +181,8 @@ public void addVertexLabel(VertexLabel vertexLabel) {
...
+            return VertexLabel.ALL_VL;
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/tx/SchemaTransaction.java:185]: OLAP_VL

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java [hunk 34]
```diff
@@ -112,11 +112,14 @@ public default void setCounterLowest(HugeType type, long lowest) {
     }
 
     public default String olapTableName(HugeType type) {
-        return this.store() + "_" + HugeType.OLAP.string() + "_" + type.string();
+        return this.store() + "_" +
+               HugeType.OLAP.string().toLowerCase() + "_" +
+               type.string().toLowerCase();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java:117]: each query will call this method?
use StringBuilder with init-capacity

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java [hunk 35]
```diff
@@ -112,11 +112,14 @@ public default void setCounterLowest(HugeType type, long lowest) {
...
     }
 
     public default String olapTableName(Id id) {
-        return this.store() + "_" + HugeType.OLAP.string() + "_" + id.asLong();
+        return this.store() + "_" +
+               HugeType.OLAP.string().toLowerCase() + "_" + id.asLong();
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java:122]: ditto

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 36]
```diff
@@ -424,23 +444,33 @@ private void mutate(Session session, BackendAction item) {
...
-                       this.olapTableName(entry.subId()) :
-                       entry.type().string();
-                this.table(name).insert(session, entry);
+                if (entry.type() == HugeType.OLAP) {
+                    table = this.table(this.olapTableName(entry.subId()));
+                    session = this.session(HugeType.OLAP);
+                } else {
+                    table = this.table(entry.type());
+                }
+                table.insert(session, entry);
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java]: move this block to line 446

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 37]
```diff
@@ -647,6 +693,18 @@ public void rollbackTx() {
         }
     }
 
+    protected RocksDBSessions db(HugeType tableType) {
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:696]: move to the line of db(String disk) method

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStoreProvider.java [hunk 38]
```diff
@@ -62,7 +62,8 @@ public String version() {
          * [1.8] #894: encode label id in string index
          * [1.9] #295: support ttl for vertex and edge
          * [1.10] #1333: support read frequency for property key
-         * [1.11] #1506: support olap properties
+         * [1.11] #1506: support olap properties and rename read frequency to
+         *               write type
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStoreProvider.java:66]: remove "rename read frequency..."?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java [hunk 39]
```diff
@@ -112,14 +112,21 @@ public default void setCounterLowest(HugeType type, long lowest) {
     }
 
     public default String olapTableName(HugeType type) {
-        return this.store() + "_" +
-               HugeType.OLAP.string().toLowerCase() + "_" +
-               type.string().toLowerCase();
+        StringBuilder sb = new StringBuilder(7);
+        sb.append(this.store())
+          .append("_")
+          .append(HugeType.OLAP.string().toLowerCase())
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java:118]: HugeType itself is LowerCase?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java [hunk 40]
```diff
@@ -112,14 +112,21 @@ public default void setCounterLowest(HugeType type, long lowest) {
...
+          .append("_")
+          .append(type.string().toLowerCase());
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java:120]: ditto

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java [hunk 41]
```diff
@@ -112,14 +112,21 @@ public default void setCounterLowest(HugeType type, long lowest) {
...
+        return sb.toString();
     }
 
     public default String olapTableName(Id id) {
-        return this.store() + "_" +
-               HugeType.OLAP.string().toLowerCase() + "_" + id.asLong();
+        StringBuilder sb = new StringBuilder(5 + 4);
+        sb.append(this.store()).append("_")
```
Comment (javeme) [inline hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java:126]: keep the same wrap-line style

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java [hunk 42]
```diff
@@ -444,36 +444,26 @@ private void mutate(Session session, BackendAction item) {
         BackendEntry entry = item.entry();
 
         RocksDBTable table;
+        if (entry.type() != HugeType.OLAP && !entry.olap()) {
```
Comment (javeme) [inline hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:447]: not the same between `entry.olap()` and `entry.type() == HugeType.OLAP`?

File: hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java (hunk 39 above)
Comment (zhoney) [reply hugegraph-core/src/main/java/com/baidu/hugegraph/backend/store/BackendStore.java:118]: no. HugeType.string() is UpperCase

File: hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java (hunk 42 above)
Comment (zhoney) [reply hugegraph-rocksdb/src/main/java/com/baidu/hugegraph/backend/store/rocksdb/RocksDBStore.java:447]: not same. `entry.type() == HugeType.OLAP` means olap vertex, but `entry.olap()` means olap vertex index

**Review by javeme — APPROVED** [review]

**Review by Linary — APPROVED** [review]

**Pull request merged**